## Checkpoint 06/26/2024
### The aim of this script is to make sure the number of samples in the minimal metadata is equal and identical to that in our final project samples

In [1]:
import pandas as pd

In [7]:
data_dir = "/Users/cheesemania/PycharmProjects/mscthesis_wrkdir/metadata"

### Merging the metadata draft and final project samples

"metadata-draft.tsv" is the "minimal_metadata" sheet in the shared-doc/dataset-curation/metadata/Metadata

"final-projects-samples" is a summary (from Sven) of our final samples after denoising step (10,055 samples).

#### Read the dataframes

In [14]:
df1 = pd.read_csv(f'{data_dir}/metadata-draft.tsv', sep='\t')
df2 = pd.read_csv(f'{data_dir}/final-projects-samples.tsv', sep='\t')

In [15]:
df1.head()

ID Bioproject ID Continent of Geolocation Country of Geolocation  \
0  CNS0407793    CNP0002114                     Asia                  China   
1  CNS0407794    CNP0002114                     Asia                  China   
2  CNS0407795    CNP0002114                     Asia                  China   
3  CNS0407796    CNP0002114                     Asia                  China   
4  CNS0407797    CNP0002114                     Asia                  China   

  Country-level Human Development Index Industrialization Host Sex  \
0                                 0.788    Industrialized  unknown   
1                                 0.788    Industrialized  unknown   
2                                 0.788    Industrialized  unknown   
3                                 0.788    Industrialized  unknown   
4                                 0.788    Industrialized  unknown   

  Host Age Category Host Age Decade Host Age Age Detail Level  
0           unknown         unknown  unknown               L2  
1           unknown         unknown  unknown               L2  
2           unknown         unknown  unknown               L2  
3           unknown         unknown  unknown               L2  
4           unknown         unknown  unknown               L2

In [16]:
df2.head()

Project ID Sample ID
0  autism-paper_35966698       F17
1  autism-paper_35966698       F28
2  autism-paper_35966698       F23
3  autism-paper_35966698       F14
4  autism-paper_35966698        F6

In [17]:
df1 = df1.rename(columns={'ID': 'Sample ID', 'Bioproject ID': 'Project ID'})
df1.head()

Sample ID  Project ID Continent of Geolocation Country of Geolocation  \
0  CNS0407793  CNP0002114                     Asia                  China   
1  CNS0407794  CNP0002114                     Asia                  China   
2  CNS0407795  CNP0002114                     Asia                  China   
3  CNS0407796  CNP0002114                     Asia                  China   
4  CNS0407797  CNP0002114                     Asia                  China   

  Country-level Human Development Index Industrialization Host Sex  \
0                                 0.788    Industrialized  unknown   
1                                 0.788    Industrialized  unknown   
2                                 0.788    Industrialized  unknown   
3                                 0.788    Industrialized  unknown   
4                                 0.788    Industrialized  unknown   

  Host Age Category Host Age Decade Host Age Age Detail Level  
0           unknown         unknown  unknown               L2  
1           unknown         unknown  unknown               L2  
2           unknown         unknown  unknown               L2  
3           unknown         unknown  unknown               L2  
4           unknown         unknown  unknown               L2

#### Check project numbers

In [25]:
# Check if all Project IDs in df1 are in df2:
print(df1['Project ID'].isin(df2['Project ID']).all())

True


In [26]:
print(df2['Project ID'].isin(df1['Project ID']).all())

False


In [20]:
len(df1['Project ID'].unique())

67

In [21]:
len(df2['Project ID'].unique())

68

In [22]:
# Find missing Project IDs in df1
missing_ids = set(df2['Project ID']) - set(df1['Project ID'])

print("Missing Project IDs in df1:", missing_ids)

# CHECK and corrected to PRJNA579121

Missing Project IDs in df1: {'PRJNA579122'}


This project id is not present in our project list. I assume it is a typo from PRJNA579121

- Corrected.

#### Check sample numbers

Based on prior work, df1 should have more samples than df2 and also include samples in df2.

In [23]:
print(df1['Sample ID'].isin(df2['Sample ID']).all())

False


In [32]:
# Find missing Sample IDs in df1
missing_ids_df1 = set(df2['Sample ID']) - set(df1['Sample ID'])

print(len(missing_ids_df1))
print("Missing Sample IDs in df1:", missing_ids_df1)

13
Missing Sample IDs in df1: {'SRR8536912', 'SRR5098762', 'SRR5098384', 'SRR9032425', 'SRR6308536', 'SRR15164412', 'SRR5098757', 'SRR5098331', 'SRR15164353', 'SRR9032428', 'SRR15169621', 'SRR8536983', 'SRR7464987'}


Why?
- For some unkown reasons, these samples may have been accidentally deleted from our general metadata table (extracted from SRA-metadata files).

In [24]:
print(df2['Sample ID'].isin(df1['Sample ID']).all())

False


In [31]:
# Find missing Sample IDs in df2
missing_ids_df2 = set(df1['Sample ID']) - set(df2['Sample ID'])

print(len(missing_ids_df2))
print("Missing Sample IDs in df2:", missing_ids_df2)

318
Missing Sample IDs in df2: {'ERR2505293', 'ERR2505116', 'ERR2505165', 'ERR2505376', 'ERR2505286', 'ERR2505277', 'ERR2505237', 'ERR2505035', 'ERR2505063', 'ERR2505179', 'ERR2505397', 'ERR2505132', 'ERR2505093', 'ERR2505271', 'ERR2505299', 'ERR2505067', 'ERR2505386', 'ERR2505189', 'ERR2505157', 'ERR2505128', 'ERR2505099', 'ERR2505120', 'ERR2505359', 'ERR2505086', 'ERR2505223', 'ERR2505056', 'ERR2505040', 'ERR2505202', 'ERR2505357', 'ERR2505191', 'ERR2505250', 'ERR2505026', 'ERR2505358', 'ERR2505085', 'ERR2505084', 'ERR2505203', 'ERR2505077', 'ERR2505174', 'ERR2505151', 'ERR2505125', 'ERR2505140', 'ERR2505127', 'ERR2505209', 'ERR2505214', 'ERR2505130', 'ERR2505051', 'ERR2505298', 'ERR2505054', 'ERR2505257', 'ERR2505270', 'ERR2505087', 'ERR2505362', 'ERR2505183', 'ERR2505287', 'ERR2505182', 'ERR2505309', 'ERR2505199', 'ERR2505097', 'ERR2505161', 'ERR2505135', 'ERR2505152', 'ERR2505377', 'ERR2505280', 'ERR2505395', 'ERR2505225', 'ERR2505126', 'ERR2505028', 'ERR2505044', 'ERR2505388', 'E

#### Check exclusion sample ids

"exclusion-ids-final.tsv" is provided by Sven. Those are the non-ITS samples he excluded when making the "Fungal-Project-Metadata" sheet in the shared-doc/dataset-curation.

"PRJEB25916_exclusion_ids.tsv" contains the 316 non-gut samples in PRJEB25916 which were not excluded in the "Fungal-Project-Metadata". 

In [34]:
df3 = pd.read_csv(f'{data_dir}/exclusion-ids-final.tsv', sep='\t')
df4 = pd.read_csv(f'{data_dir}/PRJEB25916_exclusion_ids.tsv', sep='\t')

In [35]:
df3.head()

ID
0  ERR4253103
1  ERR4253099
2  ERR4253110
3  ERR4253112
4  ERR4253122

In [36]:
df4.head()

PRJEB25916_non_gut
0         ERR2505026
1         ERR2505027
2         ERR2505028
3         ERR2505029
4         ERR2505030

In [37]:
print(df4['PRJEB25916_non_gut'].isin(df3['ID']).all())

# CHECK

True


In [45]:
# Check if the non-gut samples in PRJEB25916 are all in the list of missing samples in df2 (potential final samples)

# Convert the 'PRJEB25916_non_gut' column to a set
non_gut_sample = set(df4['PRJEB25916_non_gut'])
print(len(non_gut_sample))

316


In [46]:
print(non_gut_sample.issubset(missing_ids_df2))

# CHECK

True


In [50]:
df1.head()

Sample ID  Project ID Continent of Geolocation Country of Geolocation  \
0  CNS0407793  CNP0002114                     Asia                  China   
1  CNS0407794  CNP0002114                     Asia                  China   
2  CNS0407795  CNP0002114                     Asia                  China   
3  CNS0407796  CNP0002114                     Asia                  China   
4  CNS0407797  CNP0002114                     Asia                  China   

  Country-level Human Development Index Industrialization Host Sex  \
0                                 0.788    Industrialized  unknown   
1                                 0.788    Industrialized  unknown   
2                                 0.788    Industrialized  unknown   
3                                 0.788    Industrialized  unknown   
4                                 0.788    Industrialized  unknown   

  Host Age Category Host Age Decade Host Age Age Detail Level  
0           unknown         unknown  unknown               L2  
1           unknown         unknown  unknown               L2  
2           unknown         unknown  unknown               L2  
3           unknown         unknown  unknown               L2  
4           unknown         unknown  unknown               L2

In [47]:
# Check if any excluded sample IDs are present in df1
excluded_ids_in_df1 = df1['Sample ID'].isin(df3['ID'])

if excluded_ids_in_df1.any():
    print("Some excluded sample IDs are present in df1:", df1[excluded_ids_in_df1])
else:
    print("All excluded sample IDs are NOT present in df1.")

Some excluded sample IDs are present in df1:        Sample ID  Project ID Continent of Geolocation Country of Geolocation  \
2906  ERR2505026  PRJEB25916                   Europe                  Italy   
2907  ERR2505027  PRJEB25916                   Europe                  Italy   
2908  ERR2505028  PRJEB25916                   Europe                  Italy   
2909  ERR2505029  PRJEB25916                   Europe                  Italy   
2910  ERR2505030  PRJEB25916                   Europe                  Italy   
...          ...         ...                      ...                    ...   
3278  ERR2505398  PRJEB25916                   Europe                  Italy   
3279  ERR2505399  PRJEB25916                   Europe                  Italy   
3280  ERR2505400  PRJEB25916                   Europe                  Italy   
3281  ERR2505401  PRJEB25916                   Europe                  Italy   
3282  ERR2505402  PRJEB25916                   Europe                  Ital

In [53]:
# Filter df1 to get the rows with excluded sample ids
excluded_samples_in_df1 = df1[excluded_ids_in_df1]
# Print the exact sample ids that are excluded
print("Excluded sample IDs present in df1:", excluded_samples_in_df1['Sample ID'].tolist())
print(len(excluded_samples_in_df1['Sample ID'].tolist()))

Excluded sample IDs present in df1: ['ERR2505026', 'ERR2505027', 'ERR2505028', 'ERR2505029', 'ERR2505030', 'ERR2505031', 'ERR2505032', 'ERR2505033', 'ERR2505034', 'ERR2505035', 'ERR2505036', 'ERR2505037', 'ERR2505038', 'ERR2505039', 'ERR2505040', 'ERR2505041', 'ERR2505042', 'ERR2505043', 'ERR2505044', 'ERR2505045', 'ERR2505046', 'ERR2505047', 'ERR2505048', 'ERR2505049', 'ERR2505050', 'ERR2505051', 'ERR2505052', 'ERR2505053', 'ERR2505054', 'ERR2505055', 'ERR2505056', 'ERR2505057', 'ERR2505058', 'ERR2505059', 'ERR2505060', 'ERR2505061', 'ERR2505062', 'ERR2505063', 'ERR2505064', 'ERR2505065', 'ERR2505066', 'ERR2505067', 'ERR2505068', 'ERR2505069', 'ERR2505070', 'ERR2505071', 'ERR2505072', 'ERR2505073', 'ERR2505074', 'ERR2505075', 'ERR2505076', 'ERR2505077', 'ERR2505078', 'ERR2505079', 'ERR2505080', 'ERR2505081', 'ERR2505082', 'ERR2505083', 'ERR2505084', 'ERR2505085', 'ERR2505086', 'ERR2505087', 'ERR2505088', 'ERR2505089', 'ERR2505090', 'ERR2505091', 'ERR2505092', 'ERR2505093', 'ERR2505094

In [55]:
# Check if excluded ids in df1 are those non-gut samples of PRJEB25916
print(excluded_samples_in_df1['Sample ID'].isin(df4['PRJEB25916_non_gut']).all())

# CHECK

True


#### NEXT STEPs:
- exclude 316 non-gut samples in PRJEB25916 from df1
- add those 13 samples in df2 to df1

In [59]:
# Filter 316 samples from df1 and get a new df1
filter_ids = df1['Sample ID'].isin(df3['ID'])

new_df1 = df1[~filter_ids]
# The ~ operator negates the boolean Series, converting True to False and False to True.
# This step filters df1 to keep only the rows where filter_ids is False.

print(new_df1)

         Sample ID   Project ID Continent of Geolocation  \
0       CNS0407793   CNP0002114                     Asia   
1       CNS0407794   CNP0002114                     Asia   
2       CNS0407795   CNP0002114                     Asia   
3       CNS0407796   CNP0002114                     Asia   
4       CNS0407797   CNP0002114                     Asia   
...            ...          ...                      ...   
10455  SRR25014852  PRJNA987148                   Europe   
10456  SRR25014853  PRJNA987148                   Europe   
10457  SRR25014854  PRJNA987148                   Europe   
10458  SRR25014855  PRJNA987148                   Europe   
10459  SRR25014856  PRJNA987148                   Europe   

      Country of Geolocation Country-level Human Development Index  \
0                      China                                 0.788   
1                      China                                 0.788   
2                      China                                 0.788   

In [62]:
df1.shape

(10460, 11)

In [61]:
new_df1.shape

(10144, 11)

In [63]:
new_df1.to_csv(f'{data_dir}/minimal_metadata_updatedV1.tsv', sep='\t', index=False)

#### The 13 samples are manually added into the "minimal_metadata_updatedV1" file. Also, some more information regarding geolocation are added. Now I have the "minimal_metadata_updatedV2" file.

### Read the dataframes again

In [64]:
df_metadataV2 = pd.read_csv(f'{data_dir}/minimal_metadata_updatedV2.tsv', sep='\t')
df_final_list = pd.read_csv(f'{data_dir}/final-projects-samples.tsv', sep='\t')

In [68]:
df_metadataV2.head()

Sample ID  Project ID Continent of Geolocation Country of Geolocation  \
0  CNS0407793  CNP0002114                     Asia                  China   
1  CNS0407794  CNP0002114                     Asia                  China   
2  CNS0407795  CNP0002114                     Asia                  China   
3  CNS0407796  CNP0002114                     Asia                  China   
4  CNS0407797  CNP0002114                     Asia                  China   

  Country-level Human Development Index Industrialization Host Sex  \
0                                 0.788    Industrialized  unknown   
1                                 0.788    Industrialized  unknown   
2                                 0.788    Industrialized  unknown   
3                                 0.788    Industrialized  unknown   
4                                 0.788    Industrialized  unknown   

  Host Age Category Host Age Decade Host Age Age Detail Level  
0           unknown         unknown  unknown               L2  
1           unknown         unknown  unknown               L2  
2           unknown         unknown  unknown               L2  
3           unknown         unknown  unknown               L2  
4           unknown         unknown  unknown               L2

In [65]:
df_metadataV2.shape

(10157, 11)

In [70]:
len(df_metadataV2['Sample ID'].unique())

10057

In [73]:
id_counts = df_metadataV2['Sample ID'].value_counts()

# Filter IDs that appear more than once
non_unique_ids = id_counts[id_counts > 1].index.tolist()
print("Non-unique sample IDs:", non_unique_ids)

Non-unique sample IDs: ['SRR8536934', 'SRR8536981', 'SRR8536921', 'SRR8536920', 'SRR8536919', 'SRR8536918', 'SRR8536917', 'SRR8536916', 'SRR8536915', 'SRR8536914', 'SRR8536913', 'SRR8536962', 'SRR8536911', 'SRR8536964', 'SRR8536965', 'SRR8536966', 'SRR8536967', 'SRR8536968', 'SRR8536969', 'SRR8536970', 'SRR8536971', 'SRR8536972', 'SRR8536973', 'SRR8536974', 'SRR8536975', 'SRR8536976', 'SRR8536977', 'SRR8536978', 'SRR8536979', 'SRR8536923', 'SRR8536924', 'SRR8536925', 'SRR8536942', 'SRR8536956', 'SRR8536955', 'SRR8536954', 'SRR8536953', 'SRR8536952', 'SRR8536963', 'SRR8536951', 'SRR8536950', 'SRR8536949', 'SRR8536948', 'SRR8536945', 'SRR8536944', 'SRR8536940', 'SRR8536946', 'SRR8536939', 'SRR8536938', 'SRR8536937', 'SRR8536936', 'SRR8536961', 'SRR8536947', 'SRR8536933', 'SRR8536932', 'SRR8536929', 'SRR8536928', 'SRR8536935', 'SRR8536927', 'SRR8536980', 'SRR8536982', 'SRR8536959', 'SRR8537024', 'SRR8537014', 'SRR8537016', 'SRR8537017', 'SRR8537019', 'SRR8537020', 'SRR8537021', 'SRR853702

In [74]:
len(non_unique_ids)

100

There are 100 duplicated samples in PRJNA517994.
- excluded.

In [82]:
len(df_metadataV2['Project ID'].unique())

67

#### The 100 duplicated samples were manually deleted from the "minimal_metadata_updatedV2" file. Now I have the "minimal_metadata_updatedV3" file.

In [83]:
df_metadataV3 = pd.read_csv(f'{data_dir}/minimal_metadata_updatedV3.tsv', sep='\t')

In [84]:
df_metadataV3.shape

(10057, 11)

In [86]:
print(len(df_metadataV3['Project ID'].unique()))
print(len(df_metadataV3['Sample ID'].unique()))

67
10057


10057 is the sample number that we had before any data processing. They are raw data downloaded from various databases.

#### The final projects-samples list

In [71]:
df_final_list.head()

Project ID Sample ID
0  autism-paper_35966698       F17
1  autism-paper_35966698       F28
2  autism-paper_35966698       F23
3  autism-paper_35966698       F14
4  autism-paper_35966698        F6

In [67]:
df_final_list.shape

(10055, 2)

In [88]:
print(len(df_final_list['Project ID'].unique()))
print(len(df_final_list['Sample ID'].unique()))

67
10055


10055 is the sample number obtained from our merged feature table. It is at the point after dada2 denoising step.

### Compare our current metadata with the final list

In [89]:
# Check projects:
print(df_metadataV3['Project ID'].isin(df_final_list['Project ID']).all())
print(df_final_list['Project ID'].isin(df_metadataV3['Project ID']).all())

# GOOD

True
True


In [91]:
print(df_final_list['Sample ID'].isin(df_metadataV3['Sample ID']).all())
print(df_metadataV3['Sample ID'].isin(df_final_list['Sample ID']).all())

# metadata has 2 more samples than the final list.

True
False


In [92]:
# Find lost Sample IDs in the final list
lost_ids_metadata = set(df_metadataV3['Sample ID']) - set(df_final_list['Sample ID'])

print(len(lost_ids_metadata))
print("Lost Sample IDs in the final list:", lost_ids_metadata)

2
Lost Sample IDs in the final list: {'SRR15531435', 'SRR19047765'}


It seems that there are only 2 samples lost during our QIIME2 data processing steps.
I would say it is reasonable.

#### Manually removed these 2 samples from the metadata and now we have our "minimal_metadata_final" file！